File: ECE39595Homeworks/HW14Assigned/rkothand/Duck.java
public abstract class Duck {

   public Duck( ) { }

   public abstract void display( );
   
   public void swim( ) {
      System.out.println("All ducks float");
   }
   
   public void performQuack ( ) {
      quackBehavior.quack( );
   }
   
   public void performFly( ) {
      flyBehavior.fly( );
   }
   
   public void setFlyBehavior(FlyBehavior fb) {
      flyBehavior = fb;
   }
      
   public void setQuackBehavior(QuackBehavior qb) {
      quackBehavior = qb;
   }

   public void performEggLaying(){
      laysEggsBehavior.layEggs();
   }

   protected QuackBehavior quackBehavior;
   protected FlyBehavior flyBehavior;
   protected LaysEggs laysEggsBehavior;
}

File: ECE39595Homeworks/HW9Assigned/Unique/Duck.cpp
#include <iostream>
#include "Duck.h"
using namespace std;
Duck::Duck( ) { }

void Duck::swim( ) {
   std::cout << "All ducks float" << std::endl;
}

void Duck::performQuack ( ) {
   quackBehavior->quack( );
}

void Duck::performFly( ) {
   flyBehavior->fly( );
}

void Duck::setFlyBehavior(unique_ptr<FlyBehavior> fb) {
   flyBehavior = move(fb);
}
   
void Duck::setQuackBehavior(unique_ptr<QuackBehavior> qb) {
   quackBehavior = move(qb);
}

